server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -669,56 +683,177 @@ public String toString() {
+     * number as processed when we receive an acknowledgement for the corresponding file chunk request. With the local checkpoint tracker,
+     * we know the last acknowledged-flushed file-chunk is a file chunk whose {@code requestSeqId} equals to the local checkpoint because
+     * the recover target can flush all file chunks up to the local checkpoint.
+     * <p>
+     * When the number of un-replied file chunk requests reaches the limit (i.e. the gap between the max_seq_no and the local checkpoint is
+     * greater than {@code maxConcurrentFileChunks}), the sending thread will abort its execution. The sending process will be resumed by
+     * one of the networking threads which receive/handle the acknowledgments of the current pending file chunk requests. This process will
+     * continue until all chunks are sent and acknowledged.
+     */
+    private class MultiFileSender extends ActionRunnable<Void> implements Closeable {
+        private final Store store;
+        private final IntSupplier translogOps;
+        private final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
+        private final Semaphore semaphore = new Semaphore(0);
+        private final Iterator<StoreFileMetaData> remainingFiles;
+        private StoreFileMetaData currentFile;
+        private InputStreamIndexInput currentInput = null;
+        private long currentChunkPosition = 0;
+        private final Deque<byte[]> recycledBuffers = ConcurrentCollections.newDeque();
+        private final AtomicReference<Tuple<StoreFileMetaData, Exception>> error = new AtomicReference<>();
+
+        MultiFileSender(Store store, IntSupplier translogOps, StoreFileMetaData[] files, ActionListener<Void> listener) {
+            super(ActionListener.notifyOnce(listener));
+            this.store = store;
+            this.translogOps = translogOps;
+            this.remainingFiles = Arrays.asList(files).iterator();
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            assert ThreadPool.assertCurrentMethodIsNotCalledRecursively();
+            assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[send file chunk]");
+            while (true) {
+                assert semaphore.availablePermits() == 0;
+                if (error.get() != null) {
+                    handleErrorOnSendFiles(store, error.get().v2(), new StoreFileMetaData[]{error.get().v1()});
+                    throw error.get().v2();
+                }
+                final FileChunk chunk = readNextChunk();
+                if (chunk == null) {
+                    semaphore.release(); // allow other threads respond if we are not done yet.
+                    if (requestSeqIdTracker.getMaxSeqNo() == requestSeqIdTracker.getProcessedCheckpoint() && semaphore.tryAcquire()) {
+                        listener.onResponse(null);
```

Comment:
If this is cancelled, I think there is a potential race condition. We could have sent the last request, then discover the cancel just before leaving `cancellableThreads.execute()`. We then risk the response coming back here and we end up responding both `onResponse` and `onFailure`? Doing checkForCancel at the beginning of the while loop should rule this out.

Reply:
We wrap the listener to ensure that we notify it once. However, an explicit check is much better. I pushed https://github.com/elastic/elasticsearch/pull/44040/commits/97b209994e6d084fcbd5d7eb24ab5af41fbbb51e.